rust/connlib/tunnel/src/peer.rs

```diff
@@ -335,6 +535,87 @@ pub struct ClientOnGateway {
     ipv6: Ipv6Addr,
     resources: HashMap<ResourceId, Vec<ResourceOnGateway>>,
     filters: IpNetworkTable<FilterEngine>,
+    permanent_translations: HashMap<IpAddr, (IpAddr, ResourceId, DomainName, Instant, bool)>,
+    nat_table: NatTable,
+    buffered_events: VecDeque<GatewayEvent>,
+}
+
+#[derive(Default, Debug)]
+struct NatTable {
+    table: BiMap<(Protocol, IpAddr), (Protocol, IpAddr)>,
+    last_seen: HashMap<(Protocol, IpAddr), Instant>,
+}
+
+impl NatTable {
+    fn expire(&mut self, now: Instant) {
+        let mut removed = Vec::new();
+        for (r, e) in self.last_seen.iter() {
+            if now.duration_since(*e) >= Duration::from_secs(60) {
+                self.table.remove_by_right(r);
+                removed.push(*r);
+            }
+        }
+
+        for r in removed {
+            self.last_seen.remove(&r);
+        }
+    }
+
+    fn translate(
+        &mut self,
+        outgoing_pkt: &IpPacket,
+        real_address: IpAddr,
+        now: Instant,
+    ) -> Option<(Protocol, IpAddr)> {
+        let source_protocol = outgoing_pkt.source_protocol()?;
+        if let Some(res) = self
+            .table
+            .get_by_left(&(source_protocol, outgoing_pkt.destination()))
+        {
+            if res.1 == real_address {
+                self.last_seen.insert(*res, now);
+                return Some(*res);
+            }
+        }
+
+        let mut occupied_ports = self
+            .table
+            .iter()
+            .filter(|(_, (proto, ip))| *ip == real_address && proto.same_type(&source_protocol))
+            .map(|(_, (proto, _))| proto.value())
+            .sorted_unstable();
+
+        for p in 1.. {
+            if !occupied_ports.contains(&p) {
```

Comment:
We need to prioritise adding cases to `tunnel_test` that rely on correct port translation.

Reply:
> I actually think this doesn't work the way you want it to. I just run into the same bug today. contains() will mutate the iterator (which is why you had to use mut) and consume items on the way.

Yeah, that's why the `occupied_ports` is sorted, so that we don't consume a non-checked port.

Reply:
> We need to prioritise adding cases to tunnel_test that rely on correct port translation

This can only be tested with the client updates, I will add some basic check for that in this PR only for the nat table.

Reply:
maybe we can use a `RangeInclusiveSet` we are already using those for the filters, we build one here with all the ports then we can use [`gaps`](https://docs.rs/rangemap/latest/rangemap/inclusive_set/struct.RangeInclusiveSet.html#method.gaps) and that's it

Reply:
> > I actually think this doesn't work the way you want it to. I just run into the same bug today. contains() will mutate the iterator (which is why you had to use mut) and consume items on the way.
> 
> Yeah, that's why the `occupied_ports` is sorted, so that we don't consume a non-checked port.

Are you sure it doesn't have to consume the first item to check against it?

Reply:
when I tested this the ports where assigned sequentially as I expected, but some unit tests can reveal this quickly

Reply:
> maybe we can use a `RangeInclusiveSet` we are already using those for the filters, we build one here with all the ports then we can use [`gaps`](https://docs.rs/rangemap/latest/rangemap/inclusive_set/struct.RangeInclusiveSet.html#method.gaps) and that's it

It doesn't happen on every packet but it will happen quite often so we should use something reasonable efficient!

Reply:
> It doesn't happen on every packet but it will happen quite often so we should use something reasonable efficient!

Makes sense, I think I have an idea on how to do this in O(n) instead of O(n * log(n)) without duplicating information

Reply:
alright, pushed the changes, incorporated your previous comment too, let me know what you think